07_StagingAndMultipleQueues/main.cpp

```diff
@@ -2,78 +2,580 @@
+		m_histogramSavedSeamphore = m_device->createSemaphore(TIMELINE_SEMAPHORE_STARTING_VALUE);
+
+		// TODO: create/initialize array of atomic pointers to IGPUImage* and IGPUBuffer* to hold results
+		// no need i think
+
+		std::thread loadImagesThread(&StagingAndMultipleQueuesApp::loadImages, this);
+		std::thread saveHistogramsThread(&StagingAndMultipleQueuesApp::saveHistograms, this);
+
+		calculateHistograms();
+
+		loadImagesThread.join();
+		saveHistogramsThread.join();
+
+		return true;
+	}
+
+	//
+	void workLoopBody() override {}
+
+	//
+	bool keepRunning() override { return false; }
+
+	//
+	bool onAppTerminated() override
+	{
+		return device_base_t::onAppTerminated();
+	}
+
+protected:
+	// Override will become irrelevant in the vulkan_1_3 branch
+	SPhysicalDeviceFeatures getRequiredDeviceFeatures() const override
+	{
+		auto retval = device_base_t::getRequiredDeviceFeatures();
+		//retval.shaderStorageImageWriteWithoutFormat = true;
+		//retval.vulkanMemoryModelDeviceScope = true; // Needed for atomic operations.
+		return retval;
+	}
+
+	// Ideally don't want to have to 
+	SPhysicalDeviceFeatures getPreferredDeviceFeatures() const override
+	{
+		auto retval = device_base_t::getPreferredDeviceFeatures();
+		//retval.shaderStorageImageReadWithoutFormat = true;
+		return retval;
+	}
+
+	template<typename... Args>
+	void logFailAndTerminate(const char* msg, Args&&... args)
+	{
+		m_logger->log(msg, ILogger::ELL_ERROR, std::forward<Args>(args)...);
+		std::exit(-1);
+	}
+
+private:
+	smart_refctd_ptr<ISemaphore> m_imagesLoadedSemaphore, m_imagesProcessedSemaphore, m_histogramSavedSeamphore;
+	std::atomic<uint32_t> m_imagesLoadedCnt, m_imagesProcessedCnt, m_imagesDownloadedCnt, m_imagesSavedCnt;
+	std::array<core::smart_refctd_ptr<IGPUImage>, IMAGE_CNT> images;
+
+	static constexpr uint32_t FRAMES_IN_FLIGHT = 3u;
+	smart_refctd_ptr<video::IGPUCommandPool> commandPools[FRAMES_IN_FLIGHT];
+
+	static constexpr size_t CHANEL_CNT = 3;
+	static constexpr size_t VAL_PER_CHANEL_CNT = 256;
+	static constexpr size_t HISTOGRAM_SIZE = CHANEL_CNT * VAL_PER_CHANEL_CNT;
+	static constexpr size_t HISTOGRAM_BYTE_SIZE = HISTOGRAM_SIZE * sizeof(uint32_t);
+	static constexpr size_t COMBINED_HISTOGRAM_BUFFER_BYTE_SIZE = HISTOGRAM_BYTE_SIZE * FRAMES_IN_FLIGHT;
+	smart_refctd_ptr<IGPUBuffer> histogramBuffer = nullptr;
+	nbl::video::IDeviceMemoryAllocator::SAllocation m_histogramBufferAllocation = {};
+	// TODO: make sure ranges are ok
+	std::array<ILogicalDevice::MappedMemoryRange, FRAMES_IN_FLIGHT> m_histogramBufferMemoryRanges;
+	uint32_t* m_histogramBufferMemPtrs[3];
+
+	std::mutex assetManagerMutex;
+
+	void loadImages()
+	{
+		IAssetLoader::SAssetLoadParams lp;
+		lp.logger = m_logger.get();
+
+		// LOAD CPU IMAGES
+		core::smart_refctd_ptr<ICPUImage> cpuImages[IMAGE_CNT];
+		for(uint32_t i = 0; i < IMAGE_CNT; ++i)
+			cpuImages[i] = loadFistAssetInBundle<ICPUImage>(imagesToLoad[i]);
```

Comment:
thats actually the thing that takes the most time and RAM, move it  to the beginning of the for-loop